crates/iceberg/src/arrow/record_batch_transformer.rs

```diff
@@ -123,12 +166,55 @@ impl RecordBatchTransformer {
     pub(crate) fn build(
         snapshot_schema: Arc<IcebergSchema>,
         projected_iceberg_field_ids: &[i32],
+    ) -> Self {
+        Self::build_with_partition_data(
+            snapshot_schema,
+            projected_iceberg_field_ids,
+            None,
+            None,
+            None,
+        )
+    }
+
+    /// Build a RecordBatchTransformer with partition spec and data for proper constant identification.
+    ///
+    /// Implements the Iceberg spec's "Column Projection" rules for resolving field IDs "not present" in data files:
+    /// 1. Return the value from partition metadata if an Identity Transform exists
+    /// 2. Use schema.name-mapping.default metadata to map field id to columns without field id
+    /// 3. Return the default value if it has a defined initial-default
+    /// 4. Return null in all other cases
+    ///
+    /// # Why this method exists
+    ///
+    /// 1. **Bucket partitioning**: Distinguish identity transforms (use partition metadata constants)
+    ///    from non-identity transforms like bucket (read from data file) to enable runtime filtering on
+    ///    bucket-partitioned columns.
+    ///
+    /// 2. **Add_files field ID conflicts**: When importing Hive tables, partition columns can have field IDs
+    ///    conflicting with Parquet data columns (e.g., Parquet has field_id=1->"name", but Iceberg expects
+    ///    field_id=1->"id"). Per spec, such fields are "not present" and should use name mapping (rule #2).
+    ///
+    /// This matches Java's ParquetSchemaUtil.applyNameMapping approach but detects conflicts during projection.
+    ///
+    /// # References
+    /// - Spec: https://iceberg.apache.org/spec/#column-projection
+    /// - Java: core/src/main/java/org/apache/iceberg/util/PartitionUtil.java
+    /// - Java: parquet/src/main/java/org/apache/iceberg/parquet/ParquetSchemaUtil.java
+    pub(crate) fn build_with_partition_data(
```

Comment:
nit: How about we creat a builder struct for `RecordBatchTransformer`?

Reply:
Updated.